internal/pkg/docker/scheduler/scheduler.go

```diff
@@ -0,0 +1,312 @@
+// Scheduler functions. Errors from containers run by the scheduler
+// or from Scheduler actions are sent to the returned error channel.
+// The returned error channel is closed after calling Stop() has
+// stopped the Scheduler.
+func (s *Scheduler) Start() chan error {
+	// close done when all goroutines created by scheduler have finished
+	done := make(chan struct{})
+	errs := make(chan error)
+
+	// scheduler routine
+	s.wg.Add(1) // decremented by stopAction
+	go func() {
+		for {
+			select {
+			case action := <-s.actions:
+				if err := action.Do(s); err != nil {
+					errs <- err
+				}
+			case err := <-s.runErrs:
+				errs <- err
+			case <-done:
+				close(errs)
+				return
+			}
+		}
+	}()
+
+	go func() {
+		s.wg.Wait()
+		close(done)
+	}()
+
+	return errs
+}
+
+// RunTask stops the current running task and starts task.
+func (s *Scheduler) RunTask(task Task) {
+	// this guarantees the following:
+	// - if runTaskAction{} is pulled by the scheduler, any errors
+	//   returned by it are reported by the scheduler.
+	// - if Stop() is called _before_ the scheduler picks up this
+	//   action, then this action is skipped.
+	select {
+	case <-s.stopped:
+	case s.actions <- &runTaskAction{
+		task: task,
+	}:
+	}
+}
+
+type runTaskAction struct {
+	task Task
+}
+
+func (r *runTaskAction) Do(s *Scheduler) error {
+	// we no longer care about errors from the old task
+	taskID := s.curTaskID.Add(1)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// cancelCtxOnStop calls cancel if Stop() is called before ctx finishes.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		select {
+		case <-ctx.Done():
+		case <-s.stopped:
+			cancel()
+		}
+	}()
+
+	if taskID == 1 {
+		// start the pause container
+		opts := s.pauseRunOptions(r.task)
+		s.run(-1, opts)
+		if err := s.waitForContainerToStart(ctx, opts.ContainerName); err != nil {
+			return fmt.Errorf("wait for pause container to start: %w", err)
+		}
+	} else {
+		// ensure no pause container changes
+		curOpts := s.pauseRunOptions(s.curTask)
+		newOpts := s.pauseRunOptions(r.task)
+		if !maps.Equal(curOpts.EnvVars, newOpts.EnvVars) ||
+			!maps.Equal(curOpts.Secrets, newOpts.Secrets) ||
+			!maps.Equal(curOpts.ContainerPorts, newOpts.ContainerPorts) {
+			return errors.New("new task requires recreating pause container")
+		}
+
+		if err := s.stopTask(ctx, s.curTask); err != nil {
+			return fmt.Errorf("stop existing task: %w", err)
+		}
+	}
+
+	s.curTask = r.task
+	for name, ctr := range r.task.Containers {
+		name, ctr := name, ctr
+		s.run(taskID, s.containerRunOptions(name, ctr))
+	}
+
+	return nil
+}
+
+// Stop stops the current running task containers and the Scheduler. Stop is
+// idempotent and safe to call multiple times. Calls to RunTask() after calling Stop
+// do nothing.
+func (s *Scheduler) Stop() {
+	s.stopOnce.Do(func() {
+		close(s.stopped)
+		s.actions <- &stopAction{}
+	})
+}
+
+type stopAction struct{}
```

Comment:
❤️  how the code is arranged such that the definition of `stopAction` immediately follows the first reference of it. Very smooth to read!